### ISSUE TITLE
Searches ignored paths after "Search in Directory" was used

### ISSUE THREAD

**Issue by rm-code** [atom/find-and-replace#636]
Ok. I think it actually is a different issue altogether. 

 ```Exclude VCS Ignored Paths``` is checked in preferences:
![foo](https://cloud.githubusercontent.com/assets/11627131/12132039/44259d68-b418-11e5-973c-b1ada9956e68.png)

** Steps to reproduce **
1. Restart Atom
2. Open project and use ⇧⌘F (Find in Project) to search for a certain word
3. Close Project Find Results
4. Click on project folder and use Search in Directory
5. Close Results tab
6. Use ⇧⌘F (Find in Project)

** Results **
The first time we use ⇧⌘F it ignores the paths as expected, but after using ```Search in Directory``` all the following ⇧⌘F-Searches won't ignore paths anymore.

**Comment by lee-dohm** [comment]
* Which version of Atom is this on?
* What OS are you running?

**Comment by rm-code** [comment]
Updated description.

**Comment by amytruong** [comment]
@lee-dohm I couldn't reproduce this issue :sad: maybe you'll have better luck?

@rm-code: can you try reproducing this in safe mode? That might help a bit in finding out what's going on. Thank you!

**Comment by rm-code** [comment]
@amytruong I can reproduce it in safe mode too... but only on one project now. Here's an animation that shows how Atom responds:
![beh](https://cloud.githubusercontent.com/assets/11627131/12164379/629207c0-b515-11e5-93df-1c539c4cf16b.gif)

I can't reproduce it on the other project anymore. But it still behaves strange. When I use "Search in Project" (⇧⌘F) and afterwards try to "Search in Directory" Atom won't perform a new search. It will simply reopen the previous search results window (at least that's how it looks like) instead of searching the ignored folders as well.
![foo2](https://cloud.githubusercontent.com/assets/11627131/12164405/99502efe-b515-11e5-886c-29f4e06d5503.gif)

**Comment by adambuczynski** [comment]
I can confirm this issue as well. Sometimes, Atom ignores the ignored folders list and searches through my 10 trillion paths in the node_modules folder, almost crashing Atom in the process. I only noticed this starting to happen since Atom 1.3.x, this is on OS X El capitan.

**Comment by rm-code** [comment]
@adambuczynski Do you use Atom-Beta as well?

**Comment by adambuczynski** [comment]
No, it was too unstable for me, so I switched back to the regular atom a
while back.

**Comment by adambuczynski** [comment]
This is still an issue in the latest atom version.

`.gitignore` in root folder of project:

```
.DS_Store
node_modules
coverage
dist
*.log
```

Folder structure (simplified):

```
client/
server/
system/
  node_modules/
```

Find in files gives me results from `system/node_modules` when I specify `system` as the search location:

![image](https://cloud.githubusercontent.com/assets/490562/15266229/041b3d28-19f2-11e6-9997-97fb3bdaca20.png)

When I don't specify the `system` folder, it doesn't.

Atom version 1.9.0-dev-c4d4e0f

**Comment by lee-dohm** [comment]
@amytruong Can you reproduce this with @adambuczynski's steps?

**Comment by adambuczynski** [comment]
@lee-dohm instead of creating a new issue like you asked in atom/atom#7513, I'll just post here, since this issue pretty much describes the biggest problem find & replace currently has.

Atom version tested on: **1.9.0-dev-8beb4c2**
OS version: **OS X 10.10.5**

Steps to reproduce:

```sh
mv ~/.atom ~/.atom-backup
mkdir test
cd test
git init
echo "node_modules" >> .gitignore
echo "bower_components" >> .gitignore
mkdir client
mkdir node_modules
mkdir bower_components
mkdir client/node_modules
mkdir client/bower_components
echo "supercalifragilisticexpialidocious" > test.js
echo "supercalifragilisticexpialidocious" > client/test.js
echo "supercalifragilisticexpialidocious" > node_modules/test.js
echo "supercalifragilisticexpialidocious" > bower_components/test.js
echo "supercalifragilisticexpialidocious" > client/node_modules/test.js
echo "supercalifragilisticexpialidocious" > client/bower_components/test.js
atom . --safe
```

Resulting project view:

![image](https://cloud.githubusercontent.com/assets/490562/15374299/b1e40bb6-1d9d-11e6-8e8e-1b159db3012d.png)

Observe how tree view correctly marks all `node_modules` and `bower_components` folders as ignored, include the files therein, and including the ones in the `client` sub folder.

Now do a search in files for `supercalifragilisticexpialidocious`. Leave the folder empty:

![image](https://cloud.githubusercontent.com/assets/490562/15374335/ecd5575c-1d9d-11e6-9574-af04edf8a87e.png)

Find & replace respects the ignored folders and only shows the two relevant files in the results.
Now do the same search, but specify `client` as the target search folder:

![image](https://cloud.githubusercontent.com/assets/490562/15374366/1c1ee866-1d9e-11e6-9453-ef137fb8ea49.png)

Find & replace now ignores the rules in `.gitignore` from the parent project directory and shows the results from `bower_components` and `node_modules`.

Note that if you copy the `.gitignore` file to `client`:

```sh
cp .gitignore client
``

This has no effect:

![image](https://cloud.githubusercontent.com/assets/490562/15374555/023ae4c6-1d9f-11e6-9c4e-bdf699e5c4b5.png)

In the other issue you wrote:

> We used to go hunting up the directory structure for a Git repository if Atom couldn't find one in the root of the project. This led to serious performance concerns when people had initialized their home directory as a Git repository and then opened up directories that weren't repositories at all underneath their home directory. Because of this, we disabled support for this setup. (We also don't detect if the Git repository is in a folder that is below the project root.)

It seems that find & replace doesn't "find" any `.gitignore` if you search in a target directory.

This leads to a seriously broken find & replace functionality with very poor performance, especially if you're looking for common terms to find or replace and have a large amount of files in your `node_modules`:

![image](https://cloud.githubusercontent.com/assets/490562/15374690/995935ba-1d9f-11e6-8edd-51d4ead0b997.png)

This is unworkable as you can imagine :)

**Comment by adambuczynski** [comment]
@lee-dohm have you had a chance to look into this?

**Comment by lee-dohm** [comment]
Not as of yet. I've put it on my personal backlog.

**Comment by adambuczynski** [comment]
Cheers thanks. If you point me in the right direction I might have a look myself, see what I can find.

**Comment by adambuczynski** [comment]
@lee-dohm this issue is probably related and/or a duplicate: https://github.com/atom/find-and-replace/issues/621

**Comment by rm-code** [comment]
@adambuczynski I am not sure if it is really a duplicate, but they look related for sure. #621 as well as your https://github.com/atom/find-and-replace/issues/636#issuecomment-220155663 both seem to use the pattern field to reproduce this issue. I was getting this issue with just the normal search field, but couldn't reproduce it 100%.

**Comment by OmeGak** [comment]
Why is the status still `needs-reproduction`? This is still happening all the time in Atom 1.15. 

I have several project folders and I often do a "Search in directory" only to find I didn't explicitly exclude `node_modules` and Atom chokes a bit. Very annoying. :\

**Comment by rsese** [comment]
I was able to reproduce while poking at #890, I'll update the labels.

**Comment by 50Wliu** [comment]
Going down the rabbit hole, it seems like for the most part the paths are just passed down further and further until https://github.com/atom/scandal/blob/f11441d525106dc19ee7e5fb091b47c1ae5fedd9/src/path-filter.coffee#L50.

**Comment by 50Wliu** [comment]
Well, here we have it, it's even in the documentation.
> * `inclusions` {Array} of patterns to include. Uses minimatch with a couple additions: `['dirname']` and `['dirname/']` will match all paths in directory dirname.

Emphasis on the "`dirname` will match **all** paths in directory dirname" part.

I'll see if there's an easy way to work around this in find-and-replace.

**Comment by lee-dohm** [comment]
I'm pretty certain this was a deliberate decision based on #207 and #414.

**Comment by 50Wliu** [comment]
@lee-dohm Yeah, I'd expect that behavior if I explicitly specified `client/node_modules`, but *not* if I only specified `client`.

**Comment by lee-dohm** [comment]
Without an easy way to toggle the ignoring of files, it's impossible to gauge someone's intent when they enter a "search in" specification. Also, I'm pretty certain it has been mentioned in a few places in issues and the like that the easy way to disable the ignoring of files in project search is to specify a "search in" value.

I feel like this is one of those situations where we need to gather the various user stories and deliberately design something out rather than treat this as a bug and just make a quick change.

**Comment by sprat** [comment]
As a side note, "global" git ignores (in ~/.gitconfig) are not taken into account.

**Comment by lee-dohm** [comment]
@sprat It's already filed ... https://github.com/atom/atom/issues/12452

**Comment by sprat** [comment]
@lee-dohm not exactly the same issue !

I have a directory ignored in my global git ignore file (not in the project's .gitignore file), which doesn't appear in the project tree when "hideVCSIgnoredFile" is checked (as expected), but is still searched into when I use the Find and Replace feature. Maybe that's another aspect of this issue?

**Comment by igorpupkinable** [comment]
This still happens in v1.25.1 x64 Ubuntu.

I have 7 projects open. If I search in directory of a particular project then it searches everywhere inside that project regardless of `.gitignore` content in that project.

**Comment by alexbarton** [comment]
Any news on this?

This renders sub-directory-searches in big/nested project folders quite unusable …

**Comment by igorpupkinable** [comment]
I ended up using [atom-codesearch](https://atom.io/packages/atom-codesearch) with [ripgrep](https://github.com/BurntSushi/ripgrep/releases) instead.
